src/help/help_impl.cpp

```diff
@@ -838,27 +838,54 @@ void generate_races_sections(const config *help_cfg, section &sec, int level)
 		}
 	}
 
-	for(std::set<std::string, string_less>::iterator it = races.begin(); it != races.end(); ++it) {
+	// Support [race]help_taxonomy=, by adding all the races without that attribute
+	// first. This implementation assumes that there's only one level of nesting, but
+	// avoids a race condition in the order that races are added.
+	struct taxonomy_queue_type {
+		std::string id;
+		section content;
+	};
+	std::vector<taxonomy_queue_type> taxonomy_queue;
+
+	for(const auto& race_id : races) {
 		section race_section;
 		config section_cfg;
 
-		bool hidden = (visible_races.count(*it) == 0);
+		bool hidden = (visible_races.count(race_id) == 0);
```

Comment:
Either here or at the time the unit is first discovered, units need to make any parent races discovered too.

Reply:
Probably unnecessary, and in any case not part of this PR. The cosmetic glitch of a link to an undiscovered race is noted in #7241.

Reply:
Either needs to be resolved in this PR, or we merge this PR (because it's unblocks the reclassification of units) but log a bug that blocks 1.18. It's a trivial cosmetic glitch *until* we reach the point that lots of units have been reclassified and players are no longer encountering a unit that unlocks the race at the top of the tree. Without that race being discovered, all the subgroups can't find a visible place to attach to, and end up reaching the fallback code that puts them at the top level.

Reply:
There's now a ridiculous number of different tree-walks in this function, which is why it's still marked WIP, but visibility is propagated upwards.